### PR TITLE
Initial Skeleton for ChangeFeed Data Source

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
Open issues (will be addressed in subsequent PRs)
- Increased E2E test coverage
- Partitioning
- StartFrom configuration
- Spark streaming (Micro Batch) support
- Continous Processing support
- ContinuationState persistence

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:247]: This is one question I had from Scala, wouldn't this fall through to the `Timestamp.valueOf(value.asText())`, that was my original intention

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:21]: what is the value of `private[spark]` versus `private`? Doesn't `[spark]` share the code with anything inside the spark namespace?

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:29]: If the object is already `private` why do we need to specify the `[spark]` on the method?

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:47]: Good catch

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala:123]: When should we do this broadcast of the state? Asking to understand what's the pattern. I've seen in other places the opposite, basically wire the state into the builder.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala:31]: I don't believe so, I thought about it, but if the PK is nested, it makes it much harder.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedSpec.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedSpec.scala:37]: You can validate the schema returned, so you can verify that if Change Feed is Incremental, it matches with one and if it's FF with the other?

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:21]: scala naming convention for constants is Camel Case
please see this:
https://docs.scala-lang.org/style/naming-conventions.html

**Review by moderakh — COMMENTED** [review]
This TableProvider implementation and its hierarchy (Scan, etc) have a lot of common code with with the query Datasource. I wonder if we should use a singleton implementation of TableProvider?
why do we need to have two implementation of TableProvider?

do other Spark Connectors (cassandra, dynamodb, etc) also register a distinct TableProvider for streaming? Isn't it possible to fork at the lower part of the hierarchy rather than the root of the hierarchy?

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:247]: The None would happen is textNode would be of type TextNode but asText returns None (kind of null). 

I added the case None code because of a warning that the match did not have axhaustive cases - basically because the compiler recognizes that asText could technically return None.

if it is guaranteed that TextNode.asText would never return None/null in this case I would just use parseDateTimefromString(textNode.asText.get()) there in stead of the inner match/case.

Makes sense?

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala:31]: Agreed - will remove the TODO - came to the same conclusion later but forgot to remove the comment.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala:123]: This is the original place where the broadcast happens - just got re-arranged

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:21]: private for methods/fields means access is only possible form within the object (only when you mark private at the ctro (for .Net devs kind of class/object) level it implicitly becomes private[last-part-of-package]

So private[spark] on a method/field means inernal in .Net or package private in Java (nothing/default)

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:21]: ACK - will changes this

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:29]: because the "object" doesn't have visibility - the private above is for the ctor - all visibility is pushed to functions/methods/fields - private on the object just means the ctro is package private (private here means implicity private[spark]) - all members by default would still be public.
Anyway - given how confusing visibility scope is for Java/.Net developers when working in Scala I would strongly prefer explicitly listing the access modified - even when it is the default. Even in .Net I prefer adding it explicitly because it is not always obvious.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedSpec.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedSpec.scala:37]: ACK - will add that

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedTable.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedTable.scala:40]: ACK

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala:19]: Motivation: we will support different capabilities - Items won't support  MICRO_BATCH_READ or CONTINUOUS_READ - just BATCH_READ (and the former will drive the design even for BATCH_READ in change feed to make sure BATCH_READ reuses the code used for streaming. Changefeed won't support any writes - so the common denominator is BATCH_READ (implemented differently).
I don't think it will be helpful to use common base classes for both. But maybe we can take another look at it when we have feature-complete implementations for both - at that point it will be more obvious which part is common code and whether it is worth refactoring it so that we reduce code-duplication?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala:19]: Cassandra doesn't support MICRO_BATCH_READ or CONTINOUS_READ - just STREAMING_WRITE (so similar to our "Items" Data Source)

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala:19]: Dynamo DB doesn't support streaming at all

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala:19]: Kafka implementation is here - https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala:19]: But that one is really just similar to ChangeFeed - doesn't have query with push-down capabilities like the "Items" DataSource in Cosmos

**Review by moderakh — APPROVED** [review]
LGTM.

some questions and comments. the skeleton looks good to me, later (after feature complete) we should take a look if it makes sense to use one implementation of TableProvider or multiple.

Catalog api: databaseName, containerName can be inferred form Ident, please take a look at that comment.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedItemsDataSource.scala:19]: Thanks @FabianMeiswinkel  for the explanation. skeleton looks good to me, maybe after feature complete once we know more we can decide whether we can share code or not.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:247]: `parseDatetimeFromString` returns an Option (because the text might not be a datetime), in the case of None (not a datetime), opting for null makes sense, or throwing an exception saying the value does not represent a valid datetime. Weird I missed this compiler messages, maybe I have something turned off.

**Review by ealsur — APPROVED** [review]
LGTM, the only missing piece if it's in the scope of this PR, is the RowConverter support for the metadata

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:247]: Warnings are generated by ScalaStyle (not compiler - was unclear before). You can see it form the mvn output when running mvn clean compile from the sdk/cosmos folder

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:247]: And ACK - retruning an exception makes more sense - changed in latest iteration. Good catch!

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedSpec.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedSpec.scala:37]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala:248]: I didn't want to do the identifier -> Database/Container parsing twice. But the perf hit is probably small - so happy to change it if you think it gets much easier to read

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:21]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedTable.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosChangeFeedTable.scala]: Fixed in latest iteration

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:202]: Don't think this should return an Option, but rather the ObjectNode, otherwise the Seq on convertStructToSparkDataType would have an Option and not the actual value, and that might nor go well with the Schema?

This is used when the conversion is Json -> Spark row, so it needs to be treated as if it was an objectnode.

In that case, I believe it should return the `toString` of the metadata object node.

```
val metadata = objectNode.get(FullFidelityChangeFeedMetadataPropertyName)
        if (metadata != null && metadata.isObject) {
          metadata.asInstanceOf[ObjectNode].toString
        } else {
          null
        }
```

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:215]: What about using pattern?

getFullFidelityMetadata(objectNode) match

case Some(metadataNode) =>metadataNode.get(PreviousImagePropertyName).toString
case None => null

**Review by ealsur — APPROVED** [review]

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala]: Changed it to pattern matching - not perfect because of using many Java apis that could return null but two fewer lines and little easier to read. Line numbers could be reduced further with consequently applying fold/map/getOrElse/orNull but that becomes very hard to read - so I think this is a reasonable compromise

**Pull request merged**
